src/abi.md

```diff
@@ -1,21 +1,209 @@
 # Application Binary Interface (ABI)
 
-This section documents features that affect the ABI of the compiled output of
-a crate.
+r[abi]
 
-See *[extern functions]* for information on specifying the ABI for exporting
-functions. See *[external blocks]* for information on specifying the ABI for
-linking external libraries.
+## ABI Compatibility
+
+r[abi.compatibility]
+
+r[abi.compatibilty.type]
+Two types, `T` and `U`, can be *abi compatible*.
+
+r[abi.compatibility.equivalence]
+Two types `T` and `U` are *abi compatible* if:
+* They are the same type,
+* `U` is *abi compatible* with `T`, or
+* There exists a type `V`, such that `T` is *abi compatible* with `V` an `V` is *abi compatuble* with `U`,
+
+> [!NOTE]
+> These properties ensure that *abi compatibility* is an equivalence relation.
+
+r[abi.compatibility.integer]
+Two [integer types] are *abi compatible* if they have the same size and the same signednes
+
+> [!NOTE]
+> In particular, `usize` is *abi compatible* with `uN`, and `isize` is *abi compatible* with `iN` where `N` is the target_pointer_width.
+> Two integer types with different signedness, such as `u8` and `i8` are not *abi compatible*.
+
+```rust
+#[cfg(target_pointer_width="32")]
+fn foo(x: u32) -> u32{
+    x
+}
+#[cfg(target_pointer_width="64")]
+fn foo(x: u64) -> u64{
+    x
+}
+
+fn main(){
+    let f: fn(usize)->usize = unsafe{core::mem::transmute(foo as fn(_)->_)};
+    let x = 0usize;
+    let y = f(x);
+    assert_eq!(x,y);
+}
+```
+
+r[abi.compatibility.char]
+The type [`char`] is *abi compatible* with the type [`u32`][integer types].
+
+```rust
+fn foo(x: char) -> u32{
+    x as u32
+}
+
+fn main(){
+    let f: fn(u32)->char = unsafe{core::mem::transmute(foo as fn(_)->_)};
+    let x = b'A' as u32; // ascii character indecies are the same as Unicode character indecies
+    let y = f(x);
+    assert_eq!(y, 'A');
+}
+```
+
+r[abi.compatibility.pointer]
+Two [pointer types], `*mut T` and `*const U`, are *abi compatible* if the *metadata type*s of `T` and `U` are the same type.
+
+> [!NOTE]
+> [`core::marker::Sized`] types have a *metadata type* of `()`.
+
+> [!NOTE]
+> With transitivity, this applies regardless of the mutability of either pointer type
+
+```rust
+unsafe fn foo(x: *mut i32){
+   unsafe{x.write(5);}
+}
+
+fn main(){
+    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as unsafe fn(_))}; // Type Erase the function
+    let mut val = 0;
```

Comment:
this example is relying on `i32` being the fallback integer type when there is no other alternative being imposed, right?

(E.g., if one had written `5_i8` down below, then that ends up affecting the type inferred for `val`, and yields UB overall since now the write will be out-of-bounds, at least according to miri.)

I am wondering whether it would be better, for purposes of this example, to explicitly assign the `i32` type via `let val: i32 = 0;` and then you avoid discussion of how integer type fallback is handled in this part of the spec.

Reply:
Probably, yeah. I guess my brain was just on autopilot writing that test.